x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRoleMappingCleanupIT.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+package org.elasticsearch.upgrades;
+
+import org.elasticsearch.client.Request;
+import org.elasticsearch.client.Response;
+import org.elasticsearch.client.RestClient;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.xpack.core.security.authc.support.mapper.ExpressionRoleMapping;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import static org.elasticsearch.TransportVersions.SECURITY_ROLE_MAPPINGS_IN_CLUSTER_STATE;
+import static org.elasticsearch.xcontent.XContentFactory.jsonBuilder;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+
+public class SecurityIndexRoleMappingCleanupIT extends AbstractUpgradeTestCase {
+
+    public void testCleanupDuplicateMappings() throws Exception {
+        if (CLUSTER_TYPE == ClusterType.OLD) {
+            // If we're in a state where the same operator-defined role mappings can exist both in cluster state and the native store
+            // (SECURITY_ROLE_MAPPINGS_IN_CLUSTER_STATE transport added to security.role_mapping_cleanup feature added), create a state
+            // where the native store will need to be cleaned up
+            assumeTrue(
+                "Cleanup only needed before security.role_mapping_cleanup feature available in cluster",
+                clusterHasFeature("security.role_mapping_cleanup") == false
+            );
+            assumeTrue(
+                "If role mappings are in cluster state but cleanup has not been performed yet, create duplicated role mappings",
+                minimumTransportVersion().after(SECURITY_ROLE_MAPPINGS_IN_CLUSTER_STATE)
+            );
+            // Since the old cluster has role mappings in cluster state, but doesn't check duplicates, create duplicates
+            createNativeRoleMapping("operator_role_mapping_1", Map.of("meta", "test"));
+            createNativeRoleMapping("operator_role_mapping_2", Map.of("meta", "test"));
+            // API will return only native role mappings
+            assertAllRoleMappings(client(), "operator_role_mapping_1", "operator_role_mapping_2");
+        } else if (CLUSTER_TYPE == ClusterType.MIXED) {
+            // Create a native role mapping that doesn't conflict with anything before the migration run
+            createNativeRoleMapping("no_name_conflict", Map.of("meta", "test"));
```

Comment:
Ah whoops, my bad -- in a mixed cluster `createNativeRoleMapping("native_role_mapping" + ExpressionRoleMapping.READ_ONLY_ROLE_MAPPING_SUFFIX, Map.of("meta", "test"));` can result in a rejected request -- I think we want to move this to the `(CLUSTER_TYPE == ClusterType.OLD)` branch

Reply:
Ahh that's right! 

We can't create it there since we won't be bale to validate it. Not all bwc tests will hit the old cluster code because of the `assumeTrue` statements. I think we should skip it in this case for that reason.